new: configure syscall buffer dimension from Falco

Andreagit97 opened the pull request:
**What type of PR is this?**

/kind feature

**Any specific area of the project related to this PR?**

**What this PR does / why we need it**:

As discussed in various places (https://github.com/falcosecurity/falco/pull/2164#issuecomment-1220479171, https://github.com/falcosecurity/falco/pull/2164#issuecomment-1220792012, https://github.com/falcosecurity/falco/issues/2208)  this PR adds the possibility to change the syscall buffer dimension through the Falco configuration file.

The approach used is the one described here https://github.com/falcosecurity/falco/issues/2208#issuecomment-1248354349. Users will have an array of indexes that map to precise buffer dimensions in bytes. You can find more information in the `falco.yaml` file, but just for simplicity I report them here:

```yaml
# --- [Description]
#
# This is an index that controls the dimension of the syscall buffers.
# The syscall buffer is the shared space between Falco and its drivers where all the syscall events
# are stored.
# Falco uses a syscall buffer for every online CPU, and all these buffers share the same dimension.
# So this parameter allows you to control the size of all the buffers!
#
# --- [Usage]
#
# You can choose between different indexes: from `1` to `10` (`0` is reserved for future uses).
# Every index corresponds to a dimension in bytes:
#
# [(*), 1 MB, 2 MB, 4 MB, 8 MB, 16 MB, 32 MB, 64 MB, 128 MB, 256 MB, 512 MB]
#   ^    ^     ^     ^     ^     ^      ^      ^       ^       ^       ^
#   |    |     |     |     |     |      |      |       |       |       |
#   0    1     2     3     4     5      6      7       8       9       10
#
# As you can see the `0` index is reserved, while the index `1` corresponds to
# `1 MB` and so on.
#
# These dimensions in bytes derive from the fact that the buffer size must be:
# (1) a power of 2.
# (2) a multiple of your system_page_dimension.
# (3) greater than `2 * (system_page_dimension)`.
#
# According to these constraints is possible that sometimes you cannot use all the indexes, let's consider an
# example to better understand it:
# If you have a `page_size` of 1 MB the first available buffer size is 4 MB because 2 MB is exactly
# `2 * (system_page_size)` -> `2 * 1 MB`, but this is not enough we need more than `2 * (system_page_size)`!
# So from this example is clear that if you have a page size of 1 MB the first index that you can use is `3`.
#
# Please note: this is a very extreme case just to let you understand the mechanism, usually the page size is something
# like 4 KB so you have no problem at all and you can use all the indexes (from `1` to `10`).
#
# To check your system page size use the Falco `--page-size` command line option. The output on a system with a page
# size of 4096 Bytes (4 KB) should be the following:
#
# "Your system page size is: 4096 bytes."
#
# --- [Suggestions]
#
# Before the introduction of this param the buffer size was fixed to 8 MB (so index `4`, as you can see
# in the default value below).
# You can increase the buffer size when you face syscall drops. A size of 16 MB (so index `5`) can reduce
# syscall drops in production-heavy systems without noticeable impact. Very large buffers however could
# slow down the entire machine.
# On the other side you can try to reduce the buffer size to speed up the system, but this could
# increase the number of syscall drops!
# As a final remark consider that the buffer size is mapped twice in the process' virtual memory so a buffer of 8 MB
# will result in a 16 MB area in the process virtual memory.
# Please pay attention when you use this parameter and change it only if the default size doesn't fit your use case.

syscall_buf_size_preset: 4
```

I added also a new command line option `--page-size` to let the user easily obtain the page size of their system

If you have any suggestions I'm all ears :) 

**Which issue(s) this PR fixes**:

Fixes #2208 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
new: configure syscall buffer dimension from Falco
```

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/2214#" title="Author self-approved">Andreagit97</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)~~ [Andreagit97]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

Andreagit97 commented:
@incertum @leogr @FedeDP @jasondellaluce

leogr commented on userspace/falco/app_cmdline_options.cpp:
```diff
@@ -203,7 +203,9 @@ void cmdline_options::define()
 		("v",                             "Verbose output.", cxxopts::value(verbose)->default_value("false"))
-		("version",                       "Print version number.", cxxopts::value(print_version_info)->default_value("false"));
+		("version",                       "Print version number.", cxxopts::value(print_version_info)->default_value("false"))
+		("page-size",                     "Print the system page size used to choose the syscall buffer size.", cxxopts::value(print_page_size)->default_value("false"));
```
```suggestion
		("page-size",                     "Print the system page size (may assist choose the syscall buffer size).", cxxopts::value(print_page_size)->default_value("false"));
```

FedeDP commented on userspace/falco/app_actions/compute_syscall_buffer_size.cpp:
```diff
@@ -0,0 +1,68 @@
+	long page_size = getpagesize();
+	if(page_size <= 0)
+	{
+		return run_result::fatal("\nUnable to get the system page size through 'getpagesize()'\n");
```
So, we would break Falco on system where we are not able to fetch page size (while it worked before, right?). Ok this is an extreme case, still i'd prefer to just leave the default value and stop try to change the buffer size, instead of halting Falco.

Andreagit97 replied:
uhm yeah let's say this seems a really strange case, but yeah we can start with 8 MB in this case :)

FedeDP reviewed (commented):
Left a comment; otherwise this LGTM!

leogr commented on falco.yaml:
```diff
@@ -169,6 +169,61 @@ syscall_event_drops:
+# increase the number of syscall drops!
+# So just to conclude, change this index only if you have the necessity otherwise leave it as it is!
+
+syscall_buffer_index: 4
```
```suggestion
syscall_buffer_index: 4
```

The term `index` may be misleading, IMO. What about something like `syscall_buffer_preset` or the more verbose `syscall_buf_size_preset`? 

cc @falcosecurity/falco-maintainers

incertum replied:
Very good call @leogr!  `syscall_buf_size_preset` sounds great.

Andreagit97 replied:
changed :)

leogr reviewed (commented):
This looks great.

I've just left some suggestions (nitpicking :smile_cat:)

incertum commented on falco.yaml:
```diff
@@ -169,6 +169,61 @@ syscall_event_drops:
+#
+# Before the introduction of this param the buffer size was fixed to 8 MB (so index `4`, as you can see 
+# in the default value below).
+# Unless you are sure about what you are doing please keep this value as it is, Falco should work as
```
Perhaps adjust this section? 

```
You can increase the buffer size when you face syscall drops. The preset one size larger than the default (16MB) can reduce syscall drops in production heavy systems without noticeable impact. Very large buffers however could slow down the entire machine.
```

Btw hard-coded 16MB and didn't see any impact in terms of CPU usage or slowdown, so that would probably only happen if you go extreme ... would mention that 16MB can solve the issues some end users face and make that option sound reasonable. What do you think?

In addition, @leogr worth mentioning that `k8s` plugin can impact the machine and as a consequence also cause higher syscall drops?

Andreagit97 replied:
> Btw hard-coded 16MB and didn't see any impact in terms of CPU usage or slowdown, so that would probably only happen if you go extreme ... would mention that 16MB can solve the issues some end users face and make that option sound reasonable. What do you think?

Yeah make sense

> In addition, @leogr worth mentioning that k8s plugin can impact the machine and as a consequence also cause higher syscall drops?

make sense but probably not here :thinking:

Andreagit97 replied:
changed :)

FedeDP reviewed (dismissed):
/approve

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/2214#" title="Author self-approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/falco/pull/2214#pullrequestreview-1114095833" title="Approved">FedeDP</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)~~ [Andreagit97,FedeDP]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
LGTM label has been added.  <details>Git tree hash: 5c95ce7172deeca86625ee937bbb3f6d92d7354a</details>

jasondellaluce commented on userspace/falco/app_actions/compute_syscall_buffer_size.cpp:
```diff
@@ -0,0 +1,70 @@
+	/* Check if the chosen size is a multiple of the page size. */
+	if(chosen_size % page_size != 0)
+	{
+		return run_result::fatal("The chosen size '" + std::to_string(chosen_size) + "' is not a multiple of your system page '" + std::to_string(page_size) + "'. Please choose a greater index.\n");
```
```suggestion
		return run_result::fatal("The chosen syscall buffer size '" + std::to_string(chosen_size) + "' is not a multiple of your system page size '" + std::to_string(page_size) + "'. Please configure a greater syscall_buf_size_preset value.\n");
```
This error may be too generic for the average user. We need to be specific about _which_ size is wrong, and what specific action item can fix the issue.

This applies to the check below too.

Andreagit97 replied:
Solved, ty :)

jasondellaluce commented on userspace/falco/configuration.h:
```diff
@@ -269,6 +269,9 @@ class falco_configuration
 	uint32_t m_metadata_download_watch_freq_sec;
 
+	// Index corresponding to the syscall buffer dimension.
+	uint64_t m_syscall_buf_size_preset;
```
nit: uint64_t seems a bit excessive here

Andreagit97 replied:
changed into `uint16_t` ty :)

poiana commented:
New changes are detected. LGTM label has been removed.

Andreagit97 commented:
Rebased on actual master just to have this fix https://github.com/falcosecurity/falco/pull/2215 
I've also fixed the case in which we are not able to get the page system size in the last commit

jasondellaluce commented:
/milestone 0.33.0

poiana commented:
LGTM label has been added.  <details>Git tree hash: 450a6f8a38ed518bcd8fc9921aa2ee755470b0e0</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/2214#" title="Author self-approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/falco/pull/2214#pullrequestreview-1114095833" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/falco/pull/2214#pullrequestreview-1121560231" title="Approved">leogr</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)~~ [Andreagit97,FedeDP,leogr]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

FedeDP reviewed (approved):
/approve

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/2214#" title="Author self-approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/falco/pull/2214#pullrequestreview-1121561527" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/falco/pull/2214#pullrequestreview-1121560231" title="Approved">leogr</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)~~ [Andreagit97,FedeDP,leogr]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
